zrml/neo-swaps/src/lib.rs

```diff
@@ -0,0 +1,903 @@
+                Pools::<T>::insert(market_id, pool);
+                Self::deposit_event(Event::<T>::JoinExecuted {
+                    who: who.clone(),
+                    market_id,
+                    pool_shares_amount,
+                    amounts_in,
+                    new_liquidity_parameter,
+                });
+                Ok(())
+            })
+        }
+
+        #[require_transactional]
+        fn do_exit(
+            who: T::AccountId,
+            market_id: MarketIdOf<T>,
+            pool_shares_amount: BalanceOf<T>,
+            min_amounts_out: Vec<BalanceOf<T>>,
+        ) -> DispatchResult {
+            ensure!(pool_shares_amount != Zero::zero(), Error::<T>::ZeroAmount);
+            let _ = T::MarketCommons::market(&market_id)?;
+            Pools::<T>::try_mutate_exists(market_id, |maybe_pool| {
+                let pool =
+                    maybe_pool.as_mut().ok_or::<DispatchError>(Error::<T>::PoolNotFound.into())?;
+                ensure!(
+                    pool.liquidity_shares_manager.fees == Zero::zero(),
+                    Error::<T>::OutstandingFees
+                );
+                let ratio = bdiv(
+                    pool_shares_amount.saturated_into(),
+                    pool.liquidity_shares_manager.total_shares()?.saturated_into(),
+                )?;
+                let mut amounts_out = vec![];
+                for (&asset, &min_amount_out) in pool.assets().iter().zip(min_amounts_out.iter()) {
+                    let balance_in_pool = pool.reserve_of(&asset)?;
+                    let amount_out: BalanceOf<T> =
+                        bmul(ratio, balance_in_pool.saturated_into())?.saturated_into();
+                    amounts_out.push(amount_out);
+                    ensure!(amount_out >= min_amount_out, Error::<T>::AmountOutBelowMin);
+                    T::AssetManager::transfer(asset, &pool.account_id, &who, amount_out)?;
+                }
+                for ((_, balance), &amount_out) in pool.reserves.iter_mut().zip(amounts_out.iter())
+                {
+                    *balance = balance.saturating_sub(amount_out);
+                }
+                pool.liquidity_shares_manager.exit(&who, pool_shares_amount)?;
+                if pool.liquidity_shares_manager.total_shares()? == Zero::zero() {
+                    // FIXME We will withdraw all remaining funds (there shouldn't be any unless
+                    // we're using the native currency). The implementation is a bit defensive here.
+                    // This is an ugly hack and system should offer the option to whitelist
+                    // accounts.
+                    if pool.collateral == Asset::Ztg {
+                        let remaining = T::AssetManager::free_balance(Asset::Ztg, &pool.account_id);
+                        T::AssetManager::withdraw(Asset::Ztg, &pool.account_id, remaining)?;
+                    }
+                    *maybe_pool = None; // Delete the storage map entry.
```

Comment:
Is this really equivalent to deleting or does it still consume space afterwards?

Reply:
As far as I know, these are equivalent (note we're using `try_mutate_exists`, not try `try_mutate`):

> Mutate the item, only if an Ok value is returned. Deletes the item if mutated to a None. f will always be called with an option representing if the storage item exists (Some<V>) or if the storage item does not exist (None), independent of the QueryType. (https://paritytech.github.io/polkadot-sdk/master/frame_support/storage/trait.StorageMap.html#tymethod.try_mutate_exists)